libr/arch/p/fslsp/plugin.c

```diff
@@ -0,0 +1,254 @@
+				(r_read_ble32 (op->bytes, be) >> 2) & 3,
+				r_read_ble32 (op->bytes, be) & 3,
+				r_read_ble32 (op->bytes + 4, be),
+				r_read_ble32 (op->bytes + 8, be),
+				r_read_ble32 (op->bytes + 12, be));
+		op->type = R_ANAL_OP_TYPE_MOV;
+		op->size = 16;
+		break;
+
+	case 0x80:
+	case 0x81:
+		if (disasm)
+			r_strbuf_setf (&buf_asm, "Poll %s addr 0x%x, mask 0x%x, condition 0x%x",
+				(cmd == 0x80)? "Short": "Long",
+				r_read_ble32 (op->bytes + 4, be),
+				r_read_ble32 (op->bytes + 8, be),
+				r_read_ble32 (op->bytes + 12, be));
+		op->type = R_ANAL_OP_TYPE_NOP;
+		op->size = 16;
+		break;
+
+	case 0x82:
+		if (disasm)
+			r_strbuf_setf (&buf_asm, "Wait %d cycles",
+				r_read_ble32 (op->bytes, be) & 0xffff);
+		op->type = R_ANAL_OP_TYPE_NOP;
+		op->size = 4;
+		break;
+
+	case 0x84:
+		if (disasm)
+			r_strbuf_setf (&buf_asm, "Jump offset 0x%x",
+				r_read_ble32 (op->bytes + 4, be));
+		op->type = R_ANAL_OP_TYPE_JMP;
+		op->size = 8;
+		op->jump = r_read_ble32 (op->bytes + 4, be);
+		break;
+
+	case 0x85:
+		if (disasm)
+			r_strbuf_setf (&buf_asm, "Jump Conditional offset 0x%x, condition %x",
+				r_read_ble32 (op->bytes + 4, be),
+				r_read_ble32 (op->bytes + 8, be));
+		op->type = R_ANAL_OP_TYPE_CJMP;
+		op->size = 12;
+		op->jump = r_read_ble32 (op->bytes + 4, be);
+		break;
+
+	case 0x8F:
+		if (disasm)
+			r_strbuf_setf (&buf_asm, "CRC and Stop crc 0x%x",
+				r_read_ble32 (op->bytes + 4, be));
+		op->type = R_ANAL_OP_TYPE_TRAP;
+		op->size = 8;
+		op->jump = r_read_ble32 (op->bytes + 4, be);
+		break;
+
+	case 0xFF:
+		if (disasm)
+			r_strbuf_set (&buf_asm, "Stop");
+		op->type = R_ANAL_OP_TYPE_TRAP;
+		op->size = 8;
+		break;
+
+	default:
+		return false;
+	}
+
+	op->nopcode = 2;
+
+	if (disasm)
+		op->mnemonic = r_strbuf_tostring (&buf_asm);
+
+	return true;
+}
+
+static bool fslsp_anop(RArchSession *s, RAnalOp *op, RArchDecodeMask mask) {
+	RStrBuf buf_asm;
+	r_strbuf_init (&buf_asm);
+	const bool be = R_ARCH_CONFIG_IS_BIG_ENDIAN (s->config);
+	const bool disasm = mask & R_ARCH_OP_MASK_DISASM;
+	ut32 word = r_read_ble32 (op->bytes, be);
+	ut8 header = word >> 24;
+
+	if (header == 0x80) {
+		return fslsp_ancmd (s, op, mask);
+	} else if ((header & 0xc0) == 0) {
+		if (disasm) {
+			ut8 len = (header >> 4) & 0x03;
+			ut32 mask;
+			switch (len) {
+			case 1:
+				mask = 0x000000ff;
+				break;
+			case 3:
+				mask = 0xffffffff;
+				break;
+			default:
+				return false;
+			}
+			r_strbuf_setf (&buf_asm, "CCSR Write sys_addr 0x%07x, data 0x%x",
+				word & 0xfffffff,
+				r_read_ble32 (op->bytes + 4, be) & mask);
+		}
+		op->type = R_ANAL_OP_TYPE_MOV;
+		op->size = 8;
+	} else if (((header & 0xc0) == 0x80) && (header & 0x3c)) {
+		if (disasm)
+			r_strbuf_set (&buf_asm, "Alternate Configuration Write");
+		op->type = R_ANAL_OP_TYPE_MOV;
+		op->size = 8;
+	} else {
+		return false;
+	}
+
+	if (disasm) {
+		op->mnemonic = r_strbuf_tostring (&buf_asm);
+	}
```

Comment:
Strbuf_fini or it will leak. You can also use drain() so the returned pointer is passed to the one that calls

Reply:
U can also use strbuf in the heap with new/free actually that will be non null only when disasm is requested so you can use this to spot bad uses too and drain() will work well